lib/services/expiry/expiry.go

```diff
@@ -0,0 +1,237 @@
+	})
+	defer poll.Stop()
+
+	for {
+		lease, err := services.AcquireSemaphoreLockWithRetry(ctx, semCfg)
+		if err != nil {
+			s.Log.WarnContext(ctx, "error acquiring semaphore", "error", err)
+			continue
+		}
+
+		if err := s.processRequests(ctx); err != nil {
+			s.Log.WarnContext(ctx, "error processing access requests", "error", err)
+		}
+		lease.Stop()
+		if err := lease.Wait(); err != nil {
+			s.Log.WarnContext(ctx, "error cleaning up semaphore", "error", err)
+		}
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-poll.Next():
+		}
+	}
+}
+func (s *Service) processRequests(ctx context.Context) error {
+	requestsExpired := 0
+	nextPageStart := ""
+	for {
+		var page []*types.AccessRequestV3
+		var err error
+		// Use time at read when calculating expiry of requests.
+		readTime := s.Clock.Now()
+		page, nextPageStart, err = s.getNextPageOfAccessRequests(ctx, nextPageStart)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if len(page) == 0 {
+			return nil
+		}
+		minPageDelay := time.After(retryutils.SeventhJitter(minPageDelay))
+		for _, req := range page {
+			if !s.shouldExpire(req, readTime) {
+				continue
+			}
+			if err := s.expireRequest(ctx, req); err != nil {
+				s.Log.WarnContext(ctx, "error expiring access request", "error", err)
+				continue
+			}
+			requestsExpired++
+			if requestsExpired >= maxExpiresPerCycle {
+				return nil
+			}
+		}
+		<-minPageDelay
+	}
+}
+
+func (s *Service) getNextPageOfAccessRequests(ctx context.Context, startKey string) ([]*types.AccessRequestV3, string, error) {
+	req := &proto.ListAccessRequestsRequest{
+		Sort:       proto.AccessRequestSort_CREATED,
+		Descending: true,
+		Limit:      accessRequestPageLimit,
+		StartKey:   startKey,
+	}
+	resp, err := s.AccessPoint.ListAccessRequests(ctx, req)
+	if err != nil {
+		return nil, "", trace.Wrap(err)
+	}
+
+	return resp.AccessRequests, resp.NextKey, nil
+}
+
+func (s *Service) shouldExpire(req types.AccessRequest, readTime time.Time) bool {
+	expires := req.Expiry()
+	// Add grace period for pending access requests as expiry time may be extended on approval.
+	if req.GetState() == types.RequestState_PENDING {
+		expires = expires.Add(pendingRequestGracePeriod)
+	}
+	return readTime.After(expires)
+}
+
+func (s *Service) expireRequest(ctx context.Context, req types.AccessRequest) error {
+	expiry := req.Expiry()
+	event := &apievents.AccessRequestExpire{
+		Metadata: apievents.Metadata{
+			Type: events.AccessRequestExpireEvent,
+			Code: events.AccessRequestExpireCode,
+		},
+		ResourceMetadata: apievents.ResourceMetadata{
+			Expires: req.GetAccessExpiry(),
+		},
+		RequestID:      req.GetName(),
+		ResourceExpiry: &expiry,
+	}
+	// Emit expiry event before deletion as event we know event is expired here
+	// but the deletion may fail.
+	if err := s.Emitter.EmitAuditEvent(ctx, event); err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := s.AccessPoint.DeleteAccessRequest(ctx, req.GetName()); err != nil {
```

Comment:
We should handle NotFound error here to account for potential race conditions i.e. if something else already deleted this access request. Treat NotFound gracefully basically and log something along the lines of "access request X is already deleted". Otherwise this will return an error.